### ISSUE TITLE
Adding support for /universe/ids/ endpoint

### ISSUE THREAD

**Issue by pilot538** [Celeo/rfesi#17]
Hi, I'd like to add support for the /universe/ids POST endpoint.

I am working on a Rust application and need to support a way to convert character names into character IDs, and this endpoint seemed the correct way. 

Is this currently being worked on, or is it something I can tackle and open a PR for?

Thank you!

**Comment by Celeo** [comment]
Hi!

You're welcome to take a crack at it, or I can add it - your choice.

**Comment by pilot538** [comment]
I will give it a go, thanks!

**Comment by pilot538** [comment]
@Celeo After looking at it a bit more today, I realize I might be slightly over my head. I did not realize the /universe/ids endpoint returns it based on category (agents, alliances, characters, constellations, corporations factions, inventory_types, regions, stations, and systems), so I'm not sure the best way to parse the response. 

I would guess you would need to see what type of category it returned, and then depending on that put it in the that particular result?

**Comment by Celeo** [comment]
This is the endpoint that you POST names to, in order to resolve them to IDs, right? It looks like we just send it a list of strings and it gives back a object with optional keys that are objects with "id" and "name" pairs.

```sh
curl -X POST "https://esi.evetech.net/latest/universe/ids/?datasource=tranquility&language=en" -H  "accept: application/json" -H  "Accept-Language: en" -H  "Content-Type: application/json" -H  "Cache-Control: no-cache" -d "[  \"Celeo Servasse\",  \"Veldspar\"]"
```

```json
{
  "characters": [
    {
      "id": 91316135,
      "name": "Celeo Servasse"
    },
    {
      "id": 95223939,
      "name": "Veldspar"
    }
  ],
  "inventory_types": [
    {
      "id": 1230,
      "name": "Veldspar"
    }
  ]
}
```

**Comment by pilot538** [comment]
Correct! Do we need to do any parsing of the response, or is that totally on the implementer of the crate when they call it?

**Comment by Celeo** [comment]
Nope, I'd leave interpreting the results to the caller, since they'll have a good idea of what they're expecting.

**Comment by pilot538** [comment]
Awesome, understood. 

Another question: the `api_post!` macro seems to call `self.esi.query("GET", ...` and I do not understand why that is a `GET` versus a `POST`? 

Also, for `/universe/ids` there is only one required argument, not two. Not sure how to approach that with the `api_post!` macro. 

Thank again for all your help.

**Comment by Celeo** [comment]
> Another question: the api_post! macro seems to call self.esi.query("GET", ... and I do not understand why that is a GET versus a POST?

Oh wow, that is super weird. I'll have to check out the functions that are using that to see if they work.

**Comment by Celeo** [comment]
Yup, that GET should definitely be POST.

**Comment by pilot538** [comment]
Okay, glad I'm not completely crazy! Though, it still leaves me a bit confused on how to use the macro with a single argument for the `/universe/ids` endpoint. Any advice there would be awesome

**Comment by Celeo** [comment]
In the macros, you can omit the URL replacement section, like [in this function](https://github.com/celeo/rfesi/blob/6b22b0028651ae6eefc7e623e6688dff56e44d0b/src/groups/character.rs#L91-L99):

```rs

api_post!(
    /// Get character affiliations.
    get_affiliation,
    "post_characters_affiliation",
    RequestType::Public,
    Vec<CharacterAffiliation>,
    ,
    character_ids: &[u64],
);
```

This would make a function that POSTed to that operation_id's endpoint, with no URL substitutions, and passing in an array of `u64`s.

**Comment by Celeo** [comment]
It might help to just write a function that calls `esi.query` yourself first - that'll let you test that the call is actually working before needing to worry about macros (which are nice when they're done, but definitely difficult to deal with).

**Comment by Celeo** [comment]
If you want to keep on it, please feel free. If you don't want to fuss with it, just let me know and I'll do it. Up to you!

**Comment by pilot538** [comment]
Hi, sorry for being absent! 

I am working through a few things to improve on my poor current solution, which is:

```rust
#[derive(Debug, Deserialize)]
#[allow(missing_docs)]
pub struct Ids {
    pub characters: Option<Vec<Character>>,
}

#[derive(Debug, Deserialize)]
#[allow(missing_docs)]
pub struct Character {
    pub id: u64,
    pub name: String,
}

api_post!(
        /// Get IDs from a list of names
        get_ids,
        "post_universe_ids",
        RequestType::Public,
        Ids,
        ,
        names: &Vec<String>,
    );

```

I know there is issues (`ids` shouldn't have a member `characters` but a more general one to wrap all categories), but I'm still picking up on `serde_json` and learning. Getting there though!

**Comment by Celeo** [comment]
Looking good!

**Comment by rafaga** [comment]
Just wondering Why you don't get the categories from sde ?

**Comment by pilot538** [comment]
#25 opened a PR for this issue

**Comment by Celeo** [comment]
You've solved this in #25 - thanks!
